src/plot_api/subroutines.js

```diff
@@ -397,24 +399,151 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
 }
 
 exports.drawMainTitle = function(gd) {
+    var title = gd._fullLayout.title;
     var fullLayout = gd._fullLayout;
-
     var textAnchor = getMainTitleTextAnchor(fullLayout);
     var dy = getMainTitleDy(fullLayout);
+    var y = getMainTitleY(fullLayout, dy);
+
+    if(title.text && title.automargin) {
+        var pushMargin = needsMarginPush(gd, title);
+        if(pushMargin > 0) {
+            setDflts(title, getDflts(title)[0], getDflts(title)[1]);
+            // Recalculate these since the defaults have changed
+            dy = getMainTitleDy(fullLayout);
+            y = getMainTitleY(fullLayout, dy);
+            applyTitleAutoMargin(gd, y, pushMargin);
+        }
+    }
 
     Titles.draw(gd, 'gtitle', {
         propContainer: fullLayout,
         propName: 'title.text',
         placeholder: fullLayout._dfltTitle.plot,
         attributes: {
             x: getMainTitleX(fullLayout, textAnchor),
-            y: getMainTitleY(fullLayout, dy),
+            y: y,
             'text-anchor': textAnchor,
             dy: dy
         }
     });
 };
 
+
+function isOutsideContainer(gd, title, position, y) {
+    var plotHeight = title.yref === 'paper' ? gd._fullLayout._size.h : gd._fullLayout.height;
+    var yPosTop = Lib.isTopAnchor(title) ? y : y - titleDepth(title); // Standardize to the top of the title
+    var yPosRel = position === 'b' ? plotHeight - yPosTop : yPosTop; // Position relative to the top or bottom of plot
+    if((Lib.isTopAnchor(title) && position === 't') || Lib.isBottomAnchor(title) && position === 'b') {
+        return false;
+    } else {
+        return yPosRel < titleDepth(title);
+    }
+}
+
+
+// title.y is 1 or 0 if automargin and paper ref
+// 'auto' is not supported for either title.y or title.yanchor when automargin=true
+function getDflts(title) {
+    var titleY = title.y;
+    var titleYanchor = title.yanchor;
+    if(title.automargin && title.yref === 'paper') {
+        titleY = title.y === 0 ? 0 : 1;
+        if(title.yanchor === 'auto') {
+            titleYanchor = title.y === 0 ? 'top' : 'bottom';
+        }
+    }
+    if(title.automargin && title.yref === 'container') {
+        if(title.y === 'auto') titleY = 1;
+        if(title.yanchor === 'auto') {
+            titleYanchor = title.y < 0.5 ? 'bottom' : 'top';
+        }
+    }
+    return [titleY, titleYanchor];
+}
+
+function setDflts(title, titleY, titleYanchor) {
+    if(title.yref === 'paper' && title.y !== 0 && title.y !== 1) {
+        console.warn('title.automargin=true so resetting the supplied title.y value to 1.');
+    }
+    title.y = titleY;
+    title.yanchor = titleYanchor;
+}
+
+function titleDepth(title) {
```

Comment:
@archmoj @hannahker This is the relevant code for shape labels... happy to discuss if that would be helpful.

https://github.com/plotly/plotly.js/blob/ea8e6d11e571f9c2de7ac880bafb97e59a345adb/src/components/shapes/draw.js#L682

Reply:
@archmoj I had something like this figured out, although was struggling to reference this appropriately since I need the title to be drawn first. What would be the best way to draw the title 'offscreen' as you had suggested? 

```
var titleHeight = Drawing.bBox(d3.selectAll('.gtitle').node()).height; 
```

Reply:
> 

@hannahker Please push your WIP into a dev branch so that we could look at it

Reply:
See [here](https://github.com/plotly/plotly.js/commit/e77126c2efec72e605e6e8eea9192a79844843ea)!

Reply:
This seems to work well, although I'm still noticing that the drawn title position is offset with some of the bounding box outside of the container. This isn't an issue of margins, but rather I think the `dy` attribute (see setting [here](https://github.com/plotly/plotly.js/blob/master/src/plot_api/subroutines.js#L475)). I can see this behaviour on the `master` branch as well so this might be an existing bug. 

<img width="310" alt="Screen Shot 2023-03-10 at 10 29 10 AM" src="https://user-images.githubusercontent.com/25990216/224397286-7f79e62f-d1c3-4270-9f50-8815174e0742.png">

Reply:
I agree with you that this function 
https://github.com/plotly/plotly.js/blob/ea8e6d11e571f9c2de7ac880bafb97e59a345adb/src/plot_api/subroutines.js#L475-L486
could be the source of bug on master.
Let's try not to use it when `automargin` is set to true.